### ISSUE TITLE
latest version of puma version 2.12 doesn't work with dashing gem -v 1.3.4

### ISSUE THREAD

**Issue by trekr5** [puma/puma#735]
Hello,

I'm using the Puma gem to run a dashing dashboard. This worked fine yesterday but ever since the update today of the Puma gem, I'm now unable to get the dashboard to work as the scheduling part, of the dashing gem, doesn't work.

What do I need to do within my application to get this version of Puma to work? It is running on a Linux box with Ruby -v 1.9.3

**Comment by evanphx** [comment]
What error do you get? I don't have any way to diagnose the issue without more information.

**Comment by ball-hayden** [comment]
I'm also having issues running dashing after upgrading to puma 2.12.2. I see the following in `puma.err`:

```
/usr/local/rvm/rubies/ruby-2.2.2/lib/ruby/site_ruby/2.2.0/rubygems/core_ext/kernel_require.rb:54:in `require': cannot load such file -- dashing (LoadError)
	from /usr/local/rvm/rubies/ruby-2.2.2/lib/ruby/site_ruby/2.2.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	from config.ru:1:in `block in <module:PumaRackCompat>'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/rack/builder.rb:189:in `instance_eval'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/rack/builder.rb:189:in `initialize'
	from config.ru:in `new'
	from config.ru:in `<module:PumaRackCompat>'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/rack/builder.rb:183:in `eval'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/rack/builder.rb:183:in `new_from_string'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/rack/builder.rb:174:in `parse_file'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/configuration.rb:104:in `load_rackup'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/configuration.rb:71:in `app'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/runner.rb:123:in `app'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/runner.rb:130:in `start_server'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:215:in `worker'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:109:in `block (2 levels) in spawn_workers'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:109:in `fork'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:109:in `block in spawn_workers'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:105:in `times'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:105:in `spawn_workers'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:157:in `check_workers'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:423:in `run'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cli.rb:215:in `run'
	from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/bin/puma-wild:31:in `<main>'
```

In case it is significant, I am bundling dashing from `master` (rather than the version released to rubygems).

My puma.rb:

```
DEPLOY_NAME="sparkseat_dash"

directory "/var/www/#{DEPLOY_NAME}/current"

environment 'production'

daemonize false

pidfile    "/var/www/#{DEPLOY_NAME}/shared/tmp/puma/pid"
state_path "/var/www/#{DEPLOY_NAME}/shared/tmp/puma/state"

# threads 0, 16

bind "unix:///var/www/#{DEPLOY_NAME}/shared/tmp/puma.sock"

stdout_redirect "/var/www/#{DEPLOY_NAME}/shared/log/puma.log", "/var/www/#{DEPLOY_NAME}/shared/log/puma.err", true

# === Cluster mode ===

workers 1
prune_bundler

# === Puma control rack application ===

activate_control_app
```

**Comment by evanphx** [comment]
What does your config.ru look like? Seems like it's not loading your
Gemfile.

**Comment by ball-hayden** [comment]
```
# config.ru

require 'dashing'

configure do
  set :default_dashboard, 'default'

  helpers do
    def protected!
     # Put any authentication code you want in here.
     # This method is run before accessing any resource.
    end
  end
end

map Sinatra::Application.assets_prefix do
  run Sinatra::Application.sprockets
end

run Sinatra::Application
```

**Comment by ball-hayden** [comment]
I've done a little more digging, and I get a different error if I explicitly `Bundler.require`:

```
# config.ru

require 'rubygems'
require 'bundler'

Bundler.require

require 'dashing'

configure do
  set :default_dashboard, 'default'

  helpers do
    def protected!
     # Put any authentication code you want in here.
     # This method is run before accessing any resource.
    end
  end
end

map Sinatra::Application.assets_prefix do
  run Sinatra::Application.sprockets
end

run Sinatra::Application
```

```
config.ru:8:in `block in <module:PumaRackCompat>': undefined method `configure' for #<Puma::Rack::Builder:0x00000001b39830> (NoMethodError)
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/rack/builder.rb:189:in `instance_eval'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/rack/builder.rb:189:in `initialize'
        from config.ru:in `new'
        from config.ru:in `<module:PumaRackCompat>'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/rack/builder.rb:183:in `eval'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/rack/builder.rb:183:in `new_from_string'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/rack/builder.rb:174:in `parse_file'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/configuration.rb:104:in `load_rackup'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/configuration.rb:71:in `app'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/runner.rb:123:in `app'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/runner.rb:130:in `start_server'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:215:in `worker'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:109:in `block (2 levels) in spawn_workers'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:109:in `fork'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:109:in `block in spawn_workers'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:105:in `times'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:105:in `spawn_workers'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:157:in `check_workers'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cluster.rb:423:in `run'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/lib/puma/cli.rb:215:in `run'
        from /var/www/sparkseat_dash/shared/bundle/ruby/2.2.0/gems/puma-2.12.2/bin/puma-wild:31:in `<main>'
```

**Comment by evanphx** [comment]
This should be fixed in 2.12.3. I'm pretty sure it was because the `config.ru` wasn't loading in `TOPLEVEL_BINDING`.

**Comment by iparips** [comment]
I'm experiencing the same issue with Puma version 2.13.4 & Dashing version 1.3.4.

**Comment by pedroaxl** [comment]
@iparips As a temporary (and ugly) solution, it works if you call the methods using complete namespace:

```
Sinatra::Application.configure do
  Sinatra::Application.helpers do
    def protected!
      do_something!
    end
  end

  Sinatra::Application.get '/hello_world' do
    return "hello world!"
  end

end
```

**Comment by iparips** [comment]
@pedroaxl: Thank you for your quick response. The workaround worked for me. Are there plans to add a more long term resolution for this issue in an upcoming release?

**Comment by evanphx** [comment]
This is already fixed in `2.13`.

**Comment by evanphx** [comment]
Ok, I see what the issue is. I'll see about correcting it shortly.
